plugin/percona-pam-for-mysql/src/groups.c

```diff
@@ -41,23 +43,58 @@ struct groups_iter *groups_iter_new(const char *user_name)
   int error;
   struct groups_iter *it;
 
-  it= calloc(1, sizeof(struct groups_iter));
-  if (it == NULL)
-    return NULL;
+  if (gr_buf_size <= 0)
+  {
+    long gr_size_max, pw_size_max;
+    gr_size_max= sysconf(_SC_GETGR_R_SIZE_MAX);
+    pw_size_max= sysconf(_SC_GETPW_R_SIZE_MAX);
+    gr_buf_size= gr_size_max > pw_size_max ? gr_size_max : pw_size_max;
+  }
+
+  it= (struct groups_iter *) my_malloc(sizeof(struct groups_iter),
+                                       MYF(MY_FAE | MY_ZEROFILL));
+
+  it->buf_size= gr_buf_size;
+  if (it->buf_size <= 0)
+    it->buf_size= 1024;
+
+  it->buf= (char *) my_malloc(it->buf_size, MYF(MY_FAE));
 
-  error= getpwnam_r(user_name, &pwd, it->buf, max_nss_name_len, &pwd_result);
+  while ((error= getpwnam_r(user_name, &pwd, it->buf, it->buf_size,
+                            &pwd_result)) == ERANGE)
+  {
+    it->buf_size= it->buf_size * 2;
+    it->buf= (char *) my_realloc(it->buf, it->buf_size, MYF(MY_FAE));
+  }
   if (error != 0 || pwd_result == NULL)
   {
-    free(it);
+    fprintf(stderr, "auth_pam: Unable to obtain the passwd entry for the user "
+                    "'%s'.", user_name);
+    my_free(it->buf);
+    my_free(it);
     return NULL;
   }
 
-  it->ngroups= max_number_of_groups;
+  gr_buf_size= it->buf_size;
+
+  it->ngroups= 1024;
+  it->groups= (gid_t *) my_malloc(it->ngroups * sizeof(gid_t), MYF(MY_FAE));
   error= getgrouplist(user_name, pwd_result->pw_gid, it->groups, &it->ngroups);
   if (error == -1)
   {
-    free(it);
-    return NULL;
+    it->groups= (gid_t *) my_realloc(it->groups, it->ngroups * sizeof(gid_t),
+                                     MYF(MY_FAE));
+    error= getgrouplist(user_name, pwd_result->pw_gid, it->groups,
+                        &it->ngroups);
+    if (error == -1)
+    {
+      fprintf(stderr, "auth_pam: Unable to obtain the group access list for "
+                      "the user '%s'.", user_name);
+      my_free(it->buf);
```

Comment:
Probably, yes. It has to be impossible. But&#x2026; when I tested
`getgrouplist` on the laptop with OS X, I discovered that it behaves
very strange and not at all as described in the manual.

I have a user which is a member of 15 groups and when I pass
`ngroups=2`, `getgrouplist` returns -1 as expected, but it doesn't
return actual `ngroups`.

When I pass `ngroups` = 4, it returns 4 groups out of 15 and
indicates success!

When I pass non-existent user name and id, it returns -1.

There is always room for the race condition. What if we authenticated
user with PAM, then started to look for his groups and someone else
dropped the user at the same time. Should we crash the server?